File: backend/src/main/java/com/shikhi/learning/grading/AnswerNormalizer.java
package com.shikhi.learning.grading;

import java.util.Locale;

/**
 * Normalization pipeline for free-text grading (LLD §5): trim → collapse internal
 * whitespace → case-fold → strip trailing punctuation. Curated answer variants live in the
 * content (multiple accepted answers), so this stays intentionally simple and predictable.
 */
public final class AnswerNormalizer {

	private AnswerNormalizer() {
	}

	public static String normalize(String raw) {
		if (raw == null) {
			return "";
		}
		String collapsed = raw.trim().replaceAll("\\s+", " ").toLowerCase(Locale.ROOT);
		// Strip trailing sentence punctuation (., !, ?) — English/Bengali danda handled generally.
		return collapsed.replaceAll("[.!?।]+$", "").trim();
	}
}

File: android/app/src/main/java/com/shikhi/app/data/lesson/grading/AnswerNormalizer.kt
package com.shikhi.app.data.lesson.grading

import java.util.Locale

/**
 * Verbatim Kotlin port of the backend's `AnswerNormalizer`
 * (`backend/src/main/java/com/shikhi/learning/grading/AnswerNormalizer.java`, docs/
 * 93-offline-learning-design.md §4.3): trim → collapse internal whitespace → case-fold → strip
 * trailing sentence punctuation (`.`, `!`, `?`, the Bengali `।` danda). Curated answer variants
 * live in the content itself (multiple accepted answers per exercise), so this stays
 * intentionally simple and predictable — no stemming, no fuzzy matching.
 */
object AnswerNormalizer {

	private val WHITESPACE = Regex("\\s+")
	private val TRAILING_PUNCTUATION = Regex("[.!?।]+$")

	fun normalize(raw: String?): String {
		if (raw == null) return ""
		val collapsed = raw.trim().replace(WHITESPACE, " ").lowercase(Locale.ROOT)
		return collapsed.replace(TRAILING_PUNCTUATION, "").trim()
	}
}
